### PR TITLE
Add emacs-lisp formatting

### PR DESCRIPTION
**Pull request by elken**
Created as a draft because it doesn't _quite_ work; I've tried a couple of approaches but they either don't update correctly or require a second write.

I'm _sure_ I'm implementing it incorrectly, so this is more an RFC than an actual changeset :)

### CODE REVIEW

**Comment by dakra** [comment]
There is an example to re-indent elisp in https://github.com/radian-software/apheleia/pull/63

**Comment by elken** [comment]
> There is an example to re-indent elisp in #63

That's where I start, doesn't quite work :)

File: apheleia.el [hunk 1]
```diff
@@ -927,6 +927,7 @@ being run, for diagnostic purposes."
     (gofmt . ("gofmt"))
     (google-java-format . ("google-java-format" "-"))
     (isort . ("isort" "-"))
+    (el-indent-buffer . apheleia-indent-emacs-lisp-buffer)
```
Comment (raxod502) [inline apheleia.el:930]: To make it immediately obvious what this is for
```suggestion
    (emacs-lisp-indent . apheleia-indent-emacs-lisp-buffer)
```

File: apheleia.el [hunk 2]
```diff
@@ -998,6 +999,12 @@ rather than using this system."
              (const :tag "Name of temporary file used for output" output)))
            (function :tag "Formatter function"))))
 
+(cl-defun apheleia-indent-emacs-lisp-buffer (&key buffer formatter callback remote stdin &allow-other-keys)
```
Comment (raxod502) [inline apheleia.el:1002]: Should have docstring

File: apheleia.el [hunk 3]
```diff
@@ -998,6 +999,12 @@ rather than using this system."
...
+  (with-current-buffer buffer
+    (save-excursion
+      (funcall indent-region-function (point-min) (point-max))
+      (save-buffer 1))))
```
Comment (raxod502) [inline apheleia.el:1006]: I don't think you want to invoke `save-buffer` here.

**Comment by raxod502** [comment]
What are the issues you're facing with your current implementation? Maybe I can help advise.

**Comment by elken** [comment]
> What are the issues you're facing with your current implementation? Maybe I can help advise.

You're right that I don't want to use save-buffer, but if I don't then the buffer is left changed. If I try and use the method in #63 the callback throws a wrong arguments error

**Comment by elken** [comment]
Pushed changes, this now works in emacs but the test fails for the _slightest_ diff

```diff
diff --git a/tmp/apheleia-ft-file-EDn7yr.expected b/tmp/apheleia-ft-file-Uvnxkx.actual
index 7aeecb4..c0d102c 100644
--- a/tmp/apheleia-ft-file-EDn7yr.expected
+++ b/tmp/apheleia-ft-file-Uvnxkx.actual
@@ -1,4 +1,4 @@
 (if (and (< 3 5)
-         (= 1 1))
+        (= 1 1))
     (message "true")
   (message "false"))
```

**Comment by raxod502** [comment]
> there's an issue with tabs vs spaces

Yeah, that's just because `indent-tabs-mode` foolishly defaults to non-nil. We can disable it in file-local variables, no need to pull in anything from Doom. (That function is just for fixing up plist indentation, which is a definite improvement but just not the issue at hand here.) I've pushed a commit fixing the issue.

File: apheleia.el [hunk 4]
```diff
@@ -984,19 +985,34 @@ repository is automatically prepended to $PATH (variable
...
-             (string :tag "Argument")
-             (const :tag "Look for command in node_modules/.bin" npx)
-             (const :tag "Name of file being formatted" filepath)
-             (const :tag "Name of real file used for input" file)
-             (const :tag "Name of temporary file used for input" input)
-             (const :tag "Name of temporary file used for output" output)))
-           (function :tag "Formatter function"))))
+   :type '(alist
+           :key-type symbol
+            :value-type
```
Comment (raxod502) [inline apheleia.el:990]: Hmm, what happened to this indentation?

**Comment by elken** [comment]
Related to the comment, this makes the indentation _very_ wacky

image

Thoughts?

**Comment by elken** [comment]
Technically, this should work for any lisp. Should I add tests for common lisp and maybe scheme too?

File: test/formatters/apheleia-ft.el [hunk 5]
```diff
@@ -218,65 +218,82 @@ environment variable, defaulting to all formatters."
...
+             (append `(,(expand-file-name
+                         "scripts/formatters"
+                         (file-name-directory
+                          (file-truename
+                           ;; Borrowed with love from Magit
+                           (let ((load-suffixes '(".el")))
+                             (locate-library "apheleia"))))))
+                     exec-path)))
         (setq stdout-buffer (get-buffer-create
                              (format "*apheleia-ft-stdout-%S" formatter)))
```
Comment (raxod502) [inline test/formatters/apheleia-ft.el:230]: Why do we do this unconditionally now?

File: test/formatters/apheleia-ft.el [hunk 6]
```diff
@@ -218,65 +218,82 @@ environment variable, defaulting to all formatters."
...
+                (funcall command
+                         :buffer (current-buffer)
+                         :scratch (current-buffer)
+                         :formatter formatter
+                         :callback (lambda ()))
+                (copy-to-buffer stdout-buffer (point-min) (point-max))))
+          (progn
+
+            (with-current-buffer stdout-buffer
+              (erase-buffer))
```
Comment (raxod502) [inline test/formatters/apheleia-ft.el:247]: Why do we move this higher up now?

File: test/formatters/apheleia-ft.el (hunk 5 above)
Comment (elken) [reply test/formatters/apheleia-ft.el:230]: I didn't change this :P

**Comment by elken** [comment]
OK scratch that, I was tired when I did that.

Apart from the merge conflict I'm about to resume, are you happy with this?

**Comment by elken** [comment]
@raxod502 what's left to do here? the lint failure doesn't align with how the function lints elisp.

Think we should adjust the lint step to utilize what we define here so it dogfoods itself. If desired, I can add that to this PR

**Comment by raxod502** [comment]
I feel bad for letting this sit so long, let me see if I can fix up any remaining CI failures and merge this in.

**Pull request merged**